modules/host/network.go

```diff
@@ -85,9 +85,18 @@ func (h *Host) initNetworking(address string) (err error) {
 // threadedHandleConn handles an incoming connection to the host, typically an
 // RPC.
 func (h *Host) threadedHandleConn(conn net.Conn) {
-	// We defer the conn.Close so that the conn is still closed if `Add` returns
-	// an error.
-	defer conn.Close()
+	// Close the conn immediately on host.Close.
+	defer func() {
+		// We defer the conn.Close so that the conn is still closed if `Add` returns
+		// an error.
+		defer conn.Close()
+		err := h.tg.Add()
+		if err != nil {
+			return
+		}
+		defer h.tg.Done()
+		<-h.tg.StopChan()
+	}()
```

Comment:
why is this function called with `defer` instead of `go`?
This is a tricky problem, because you don't want the goroutine to linger if the connection closes gracefully. The easiest approach I can see is:
```go
go func() {
    <-h.tg.StopChan()
    conn.Close()
}()
```
This takes advantage of the fact that `net.Conn.Close` can be safely called twice.